### PR TITLE
CA-273775: Fix race in VM_receive_memory, and improve error handling

### PR DESCRIPTION
**Pull request by mg12**
This patch series fixes two issues:

1) in case the VM_receive_memory thread raises an exception either in the VM_create or VM_restore operations, it should still kill the corresponding receive_vgpu thread.
2) there's a race in which the VM_receive_memory thread does not wait for the vgpu_receiver_sync table to be initialised in the receive_vgpu thread counterpart, also in the receiving host. The signal from the receiving host's receive_vgpu thread that the vgpu_receiver_sync table is already initialised reaches the sending host via the 'Synchronisation point 1-vgpu', but there's no signal from the sending host to the receiving host's VM_receive_memory thread. We add this missing signal from the sending host to the receiving host's VM_receivie_memory thread and call it 'Synchronisation point 1-vgpu ACK', and make sure that this ACK message is always processed by both sending and receiving hosts in the case a VGPU migration is used.

Devtests for VM.migration passed both without and with VGPU. With VGPU migration, we see in the logs:
- sender:
```
|10|Async.VM.pool_migrate R:084d7c004377|xenops_server] VM.migrate: Synchronisation point 1-vgpu
|10|Async.VM.pool_migrate R:084d7c004377|xenops_server] VM.migrate: Synchronisation point 1-vgpu ACK
|10|Async.VM.pool_migrate R:084d7c004377|xenops_server] VM.migrate: Synchronisation point 1
|10|Async.VM.pool_migrate R:084d7c004377|xenops_server] VM.migrate: Synchronisation point 2
|10|Async.VM.pool_migrate R:084d7c004377|xenops_server] VM.migrate: Synchronisation point 3
|10|Async.VM.pool_migrate R:084d7c004377|xenops_server] VM.migrate: Synchronisation point 4
```
- receiver:
```
|66|Async.VM.pool_migrate R:084d7c004377|xenops_server] VM.receive_vgpu: Synchronisation point 1-vgpu
|20|Async.VM.pool_migrate R:084d7c004377|xenops_server] VM.receive_memory: Synchronisation point 1-vgpu ACK
|20|Async.VM.pool_migrate R:084d7c004377|xenops_server] VM.receive_memory: Synchronisation point 1
|20|Async.VM.pool_migrate R:084d7c004377|xenops_server] VM.receive_memory: Synchronisation point 2
|66|Async.VM.pool_migrate R:084d7c004377|xenops_server] VM.receive_vgpu: Synchronisation point 2-vgpu
|20|Async.VM.pool_migrate R:084d7c004377|xenops_server] VM.receive_memory: Synchronisation point 3
|20|Async.VM.pool_migrate R:084d7c004377|xenops_server] VM.receive_memory: Synchronisation point 4
```

### CODE REVIEW

**Comment by mg12** [comment]
@robhoes

File: lib/xenops_server.ml [hunk 1]
```diff
@@ -1680,29 +1687,49 @@ and perform ?subtask ?result (op: operation) (t: Xenops_task.task_handle) : unit
...
-           simplify [VM_create (id, Some memory_limit);] @
-           (* Perform as many operations as possible on the destination domain before pausing the original domain *)
-           (atomics_of_operation (VM_restore_vifs id))
-         ) t;
-         Handshake.send s Handshake.Success
-       with e ->
-         Handshake.send s (Handshake.Error (Printexc.to_string e));
-         raise e
+
+      finally (fun ()->
```
Comment (mseri) [inline lib/xenops_server.ml]: `finally (finally () () ... )()` gets very hard to parse. Can you extract the needed functions and shink the double finally to a few lines?

Comment (mg12) [reply lib/xenops_server.ml]: I've removed the internal finally by moving the ACK event up, so that the ACK is always executed regardless of any failure during VM_create. This simplifies the code and we don't need the 3rd commit anymore: 905f9ad882e403eda88fffe0399dd7a7b9d8078d

**Comment by mseri** [comment]
Looks good to me, but I think @robhoes should say the final word

File: lib/xenops_server.ml [hunk 2]
```diff
@@ -1680,29 +1682,49 @@ and perform ?subtask ?result (op: operation) (t: Xenops_task.task_handle) : unit
...
+           Handshake.recv_success s;
+           debug "VM.receive_memory: Synchronisation point 1-vgpu ACK";
+           (* After this point, vgpu_receiver_sync is initialised by the corresponding receive_vgpu thread
+              and therefore can be used by this VM_receive_memory thread
+            *)
+         end
+        );
+
+        (try
+           perform_atomics (
```
Comment (lindig) [inline lib/xenops_server.ml]: Just an idea
```
List.concat 
[ simplify [VM_create (id, Some memory_limit)]
; atomics_of_operation (VM_restore_vifs id)
] 
|> fun atoms -> perform atoms t;
```

File: lib/xenops_server.ml [hunk 3]
```diff
@@ -1680,29 +1682,49 @@ and perform ?subtask ?result (op: operation) (t: Xenops_task.task_handle) : unit
...
+         with e ->
+           Handshake.send s (Handshake.Error (Printexc.to_string e));
+           raise e
+        );
+        debug "VM.receive_memory: Synchronisation point 1";
+
+        debug "VM.receive_memory restoring VM";
+        (* Check if there is a separate vGPU data channel *)
+        let vgpu_info = Stdext.Opt.of_exception (fun () -> Hashtbl.find vgpu_receiver_sync id) in
+        perform_atomics (
```
Comment (lindig) [inline lib/xenops_server.ml]: Not sure this is better:
```
VGPU_DB.ids id
|> List.map (fun vgpu_id -> VGPU_start (vgpu_id, true))
|> fun x -> x @ [VM_restore (id, FD s, Opt.map (fun x -> FD x.vgpu_fd) vgpu_info)]
|> fun atoms -> perform_atomics atoms t 
```
